oaipmh/src/main/java/whelk/export/servlet/Helpers.java

```diff
@@ -1,21 +1,176 @@
-import java.sql.PreparedStatement;
-import java.sql.SQLException;
-import java.sql.Timestamp;
+import java.io.IOException;
+import java.sql.*;
 import java.time.ZonedDateTime;
-import java.util.Arrays;
-import java.util.Enumeration;
-import java.util.HashSet;
-import java.util.StringJoiner;
+import java.util.*;
 
 public class Helpers
 {
+    public static class ResultIterator implements Iterator<Document>, AutoCloseable {
+
+        private final PreparedStatement statement;
+        private final ResultSet resultSet;
+        private final SetSpec setSpec;
+        private final boolean includeDependenciesInTimeInterval;
+        private final Stack<Document> resultingDocuments = new Stack<>();
+        private boolean firstAccess = true;
+
+        public ResultIterator(PreparedStatement statement, SetSpec setSpec, boolean includeDependenciesInTimeInterval)
+                throws SQLException {
+            this.statement = statement;
+            this.resultSet = statement.executeQuery();
+            this.setSpec = setSpec;
+            this.includeDependenciesInTimeInterval = includeDependenciesInTimeInterval;
+        }
+
+        private boolean isHeld(Document doc, String bySigel)
+        {
+            if (bySigel == null)
+                return false;
+
+            List<Document> holdings = OaiPmh.s_whelk.getStorage().getAttachedHoldings(doc.getThingIdentifiers(), OaiPmh.s_whelk.getJsonld());
+            for (Document holding : holdings)
+            {
+                String sigel = holding.getSigel();
+                if (bySigel.equals(sigel))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void emitAffected(Document updated)
+        {
+            String updatedCollection = LegacyIntegrationTools.determineLegacyCollection(updated, OaiPmh.s_whelk.getJsonld());
+
+            if (updatedCollection == null)
+            {
+                return;
+            }
+
+            if (setSpec == null || setSpec.getRootSet() == null)
+            {
+                // If no set is used, all records are welcome.
+                resultingDocuments.push(updated);
+            }
+            else if (updatedCollection.equals("auth"))
+            {
+                if (setSpec.getRootSet().equals("auth"))
+                    resultingDocuments.push(updated);
+                if (includeDependenciesInTimeInterval && (setSpec.getRootSet().equals("bib") || setSpec.getRootSet().equals("hold")))
+                {
+                    List<Tuple2<String, String>> dependers = OaiPmh.s_whelk.getStorage().followDependers(updated.getShortId(), JsonLd.getNON_DEPENDANT_RELATIONS());
+                    for (Tuple2<String, String> depender : dependers)
+                    {
+                        String dependerId = depender.getFirst();
+                        Document dependerDocument = OaiPmh.s_whelk.loadEmbellished(dependerId);
```

Comment:
Should this be `getDocument()`? (embellish is done in `ResponseCommon.emitRecord()`)

Reply:
Good catch! Fixing it!